In [1]:
import pandas as pd

In [2]:
df=pd.read_csv(r'company.csv')

In [3]:
df.head()

Fiscal Year  Total Revenue  Net Income  Total Assets  Total Liabilities  \
0         2021         168088       61271        333779             191791   
1         2022         198270       72738        364840             198298   
2         2023         211915       72361        411976             205753   
3         2021          53823        5519         62131              30548   
4         2022          81462       12556         82338              36440   

   Cash Flow from Operating Activities    Company  
0                                76740  Microsoft  
1                                89035  Microsoft  
2                                87582  Microsoft  
3                                11497      Tesla  
4                                14724      Tesla

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9 entries, 0 to 8
Data columns (total 7 columns):
 #   Column                               Non-Null Count  Dtype 
---  ------                               --------------  ----- 
 0   Fiscal Year                          9 non-null      int64 
 1   Total Revenue                        9 non-null      int64 
 2   Net Income                           9 non-null      int64 
 3   Total Assets                         9 non-null      int64 
 4   Total Liabilities                    9 non-null      int64 
 5   Cash Flow from Operating Activities  9 non-null      int64 
 6   Company                              9 non-null      object
dtypes: int64(6), object(1)
memory usage: 632.0+ bytes


In [5]:
df.describe()

Fiscal Year  Total Revenue    Net Income   Total Assets  \
count     9.000000       9.000000      9.000000       9.000000   
mean   2022.000000  217084.555556  58988.555556  268669.111111   
std       0.866025  133913.789489  38258.207672  140780.645366   
min    2021.000000   53823.000000   5519.000000   62131.000000   
25%    2021.000000   96773.000000  14974.000000  106618.000000   
50%    2022.000000  198270.000000  72361.000000  351002.000000   
75%    2023.000000  365817.000000  94680.000000  352755.000000   
max    2023.000000  394328.000000  99803.000000  411976.000000   

       Total Liabilities  Cash Flow from Operating Activities  
count           9.000000                             9.000000  
mean       176252.333333                         69951.777778  
std        112621.131472                         44632.990091  
min         30548.000000                         11497.000000  
25%         43009.000000                         14724.000000  
50%        198298.000000                         87582.000000  
75%        287912.000000                        104038.000000  
max        302083.000000                        122151.000000

In [6]:
df['Revenue Growth (%)'] = df.groupby(['Company'])['Total Revenue'].pct_change() * 100
df['Net Income Growth (%)'] = df.groupby(['Company'])['Net Income'].pct_change() * 100
df['Assets Growth (%)'] = df.groupby(['Company'])['Total Assets'].pct_change() * 100
df['Liabilities Growth (%)'] = df.groupby(['Company'])['Total Liabilities'].pct_change() * 100
df['Cash Flow Growth (%)'] = df.groupby(['Company'])['Cash Flow from Operating Activities'].pct_change() * 100


In [7]:
df.head(9)

Fiscal Year  Total Revenue  Net Income  Total Assets  Total Liabilities  \
0         2021         168088       61271        333779             191791   
1         2022         198270       72738        364840             198298   
2         2023         211915       72361        411976             205753   
3         2021          53823        5519         62131              30548   
4         2022          81462       12556         82338              36440   
5         2023          96773       14974        106618              43009   
6         2021         365817       94680        351002             287912   
7         2022         394328       99803        352755             302083   
8         2023         383285       96995        352583             290437   

   Cash Flow from Operating Activities    Company  Revenue Growth (%)  \
0                                76740  Microsoft                 NaN   
1                                89035  Microsoft           17.956071   
2                                87582  Microsoft            6.882030   
3                                11497      Tesla                 NaN   
4                                14724      Tesla           51.351653   
5                                13256      Tesla           18.795267   
6                               104038      Apple                 NaN   
7                               122151      Apple            7.793788   
8                               110543      Apple           -2.800461   

   Net Income Growth (%)  Assets Growth (%)  Liabilities Growth (%)  \
0                    NaN                NaN                     NaN   
1              18.715216           9.305858                3.392756   
2              -0.518299          12.919636                3.759493   
3                    NaN                NaN                     NaN   
4             127.504983          32.523217               19.287678   
5              19.257725          29.488207               18.026894   
6                    NaN                NaN                     NaN   
7               5.410858           0.499427                4.921990   
8              -2.813543          -0.048759               -3.855232   

   Cash Flow Growth (%)  
0                   NaN  
1             16.021631  
2             -1.631942  
3                   NaN  
4             28.068192  
5             -9.970117  
6                   NaN  
7             17.409985  
8             -9.502992

In [8]:
df.fillna(method='bfill', inplace=True)

In [9]:
df.head(9)

Fiscal Year  Total Revenue  Net Income  Total Assets  Total Liabilities  \
0         2021         168088       61271        333779             191791   
1         2022         198270       72738        364840             198298   
2         2023         211915       72361        411976             205753   
3         2021          53823        5519         62131              30548   
4         2022          81462       12556         82338              36440   
5         2023          96773       14974        106618              43009   
6         2021         365817       94680        351002             287912   
7         2022         394328       99803        352755             302083   
8         2023         383285       96995        352583             290437   

   Cash Flow from Operating Activities    Company  Revenue Growth (%)  \
0                                76740  Microsoft           17.956071   
1                                89035  Microsoft           17.956071   
2                                87582  Microsoft            6.882030   
3                                11497      Tesla           51.351653   
4                                14724      Tesla           51.351653   
5                                13256      Tesla           18.795267   
6                               104038      Apple            7.793788   
7                               122151      Apple            7.793788   
8                               110543      Apple           -2.800461   

   Net Income Growth (%)  Assets Growth (%)  Liabilities Growth (%)  \
0              18.715216           9.305858                3.392756   
1              18.715216           9.305858                3.392756   
2              -0.518299          12.919636                3.759493   
3             127.504983          32.523217               19.287678   
4             127.504983          32.523217               19.287678   
5              19.257725          29.488207               18.026894   
6               5.410858           0.499427                4.921990   
7               5.410858           0.499427                4.921990   
8              -2.813543          -0.048759               -3.855232   

   Cash Flow Growth (%)  
0             16.021631  
1             16.021631  
2             -1.631942  
3             28.068192  
4             28.068192  
5             -9.970117  
6             17.409985  
7             17.409985  
8             -9.502992

In [10]:
growth_aggregated_df = df.groupby(['Company', 'Fiscal Year']).agg(
    {
        'Revenue Growth (%)': 'mean',
        'Net Income Growth (%)': 'mean',
        'Assets Growth (%)': 'mean',
        'Liabilities Growth (%)': 'mean',
        'Cash Flow Growth (%)': 'mean'
    }
).reset_index()

# Rename columns for clarity
growth_aggregated_df.columns = [
    'Company', 'Fiscal Year', 
    'Avg Revenue Growth (%)', 
    'Avg Net Income Growth (%)', 
    'Avg Assets Growth (%)', 
    'Avg Liabilities Growth (%)', 
    'Avg Cash Flow Growth (%)'
]

In [11]:
growth_aggregated_df.head(9)

Company  Fiscal Year  Avg Revenue Growth (%)  Avg Net Income Growth (%)  \
0      Apple         2021                7.793788                   5.410858   
1      Apple         2022                7.793788                   5.410858   
2      Apple         2023               -2.800461                  -2.813543   
3  Microsoft         2021               17.956071                  18.715216   
4  Microsoft         2022               17.956071                  18.715216   
5  Microsoft         2023                6.882030                  -0.518299   
6      Tesla         2021               51.351653                 127.504983   
7      Tesla         2022               51.351653                 127.504983   
8      Tesla         2023               18.795267                  19.257725   

   Avg Assets Growth (%)  Avg Liabilities Growth (%)  Avg Cash Flow Growth (%)  
0               0.499427                    4.921990                 17.409985  
1               0.499427                    4.921990                 17.409985  
2              -0.048759                   -3.855232                 -9.502992  
3               9.305858                    3.392756                 16.021631  
4               9.305858                    3.392756                 16.021631  
5              12.919636                    3.759493                 -1.631942  
6              32.523217                   19.287678                 28.068192  
7              32.523217                   19.287678                 28.068192  
8              29.488207                   18.026894                 -9.970117

### **Summary of Findings**

The analysis of year-over-year growth across key financial metrics for three major companies—Apple, Microsoft, and Tesla—reveals distinct patterns in their financial performance from 2021 to 2023.

1. **Apple**:
   - **Revenue and Net Income**: Apple showed moderate growth in revenue and net income in 2021 and 2022, with an average growth rate of approximately 7.8% and 5.4%, respectively. However, in 2023, both revenue and net income declined by around 2.8%. This suggests a slowdown in Apple's financial performance, possibly due to market saturation, economic conditions, or competitive pressures.
  
  - **Assets and Liabilities**: The company's assets and liabilities remained relatively stable, with slight growth in 2021 and 2022, followed by a small decline in 2023. This could indicate careful management of its financial structure, though the decline in assets and liabilities in 2023 could be a sign of decreased investment or asset disposal.
  
  - **Cash Flow**: Cash flow from operating activities saw significant growth in 2021 and 2022 but dropped sharply by 9.5% in 2023, reflecting potential challenges in maintaining operating efficiency or changes in working capital requirements.

2. **Microsoft**:
   - **Revenue**: Microsoft experienced strong revenue growth in 2021 and 2022, with an average increase of about 18%. However, the growth rate dropped to 6.9% in 2023, indicating a potential deceleration in sales or other income-generating activities.
   - **Net Income**: The net income growth mirrored the revenue trend, with a significant increase in 2021 and 2022, followed by a minor decline in 2023. This suggests that Microsoft's profitability was slightly impacted in 2023 despite continued revenue growth.
   - **Assets and Liabilities**: Microsoft’s assets grew steadily across the years, with the highest growth in 2023 (12.9%), possibly due to acquisitions or new investments. Liabilities also increased but at a slower rate, indicating a strengthening balance sheet.
   - **Cash Flow**: Cash flow from operating activities remained strong in 2021 and 2022, but there was a slight negative growth in 2023 (-1.6%), possibly indicating tighter cash flows despite increased assets.

3. **Tesla**:
   - **Revenue**: Tesla exhibited remarkable revenue growth in 2021 and 2022, with a staggering 51.4% increase each year. This reflects Tesla's rapid expansion and increased market penetration during this period. However, the growth rate moderated to 18.8% in 2023, which, while still strong, shows signs of maturing growth.
   - **Net Income**: Tesla’s net income saw an extraordinary growth of over 127% in 2021 and 2022, reflecting improved profitability. This dropped to 19.3% in 2023, which is still impressive but much lower than the previous years.
   - **Assets and Liabilities**: Tesla’s assets grew significantly, especially in 2021 and 2022, indicating heavy investment in production capacity and other assets. Liabilities also grew but were well-managed relative to the increase in assets.
   - **Cash Flow**: Tesla's cash flow from operating activities grew strongly in 2021 and 2022, consistent with its revenue and net income growth. However, in 2023, it decreased by almost 10%, possibly due to increased capital expenditures or changes in working capital requirements.

### **Conclusion**
- **Growth Slowdown**: Both Apple and Microsoft experienced a slowdown in revenue and net income growth in 2023, whereas Tesla, though still growing, also showed signs of deceleration. This trend could indicate broader economic factors affecting these companies.
- **Investment and Assets**: Microsoft and Tesla continued to grow their assets, with Tesla showing particularly strong asset growth, indicating continued investment in infrastructure and capacity.
- **Cash Flow Challenges**: All three companies saw declines in cash flow growth in 2023, suggesting potential challenges in managing operating cash flows despite overall growth in previous years.

